src/decoder.rs

```diff
@@ -1,69 +1,80 @@
-use DecodeError;
 use bigint::BigUint;
+use DecodeError;
+
+pub(crate) trait Decoder<'a>: Copy {
+    type Iter: std::iter::Iterator;
+    fn iter(&'a str) -> Self::Iter;
+    fn first(&[Self]) -> <Self::Iter as std::iter::Iterator>::Item;
+    fn carry(&[Self], <Self::Iter as std::iter::Iterator>::Item) -> Result<u32, DecodeError>;
+}
+impl<'a> Decoder<'a> for char {
+    type Iter = std::str::Chars<'a>;
+    fn iter(s: &'a str) -> Self::Iter {
+        s.chars()
+    }
+    fn first(this: &[Self]) -> <Self::Iter as std::iter::Iterator>::Item {
+        this[0]
+    }
+    fn carry(this: &[Self], c: char) -> Result<u32, DecodeError> {
+        this.iter()
+            .enumerate()
+            .find(|&(_, ch)| *ch == c)
+            .map(|(i, _)| i as u32)
+            .ok_or(DecodeError)
+    }
+}
+const INVALID_INDEX: u8 = 0xFF;
 
-pub struct AsciiDecoder;
-pub struct Utf8Decoder;
+impl<'a> Decoder<'a> for u8 {
+    type Iter = std::str::Bytes<'a>;
+    fn iter(s: &'a str) -> Self::Iter {
+        s.bytes()
+    }
+    fn first(this: &[Self]) -> <Self::Iter as std::iter::Iterator>::Item {
+        this[0]
+    }
+    fn carry(this: &[Self], c: u8) -> Result<u32, DecodeError> {
+        let mut lookup = [INVALID_INDEX; 256];
 
-macro_rules! decode {
-    ($alpha:ident, $input:ident, $iter:ident, $c:pat => $carry:expr) => ({
-        if $input.len() == 0 {
-            return Ok(Vec::new());
+        for (i, byte) in this.iter().enumerate() {
```

Comment:
I think this is where the performance hit is from. The lookup table was previously computed once per decode, while now it is being re-computed for every single byte of input.

There might be a small hit due to the fact that a static code injection with early return has been replaced by a function call with error handling (my experience is that LLVM isn't always so good at optimizing it, and it's worse for `?` than it is for `try!` in some cases), but there is no way around this one if we want to use traits instead of macros.

Reply:
Oh! I haven't realized... The biggest performance hit was from my introducing a bug in the `BigInt::div_mod` function, when I was refactoring it.

Reply:
Happens to the best of us